include/message_filters/sync_policies/latest_time.h

```diff
@@ -0,0 +1,276 @@
+struct LatestTime : public PolicyBase<M0, M1, M2, M3, M4, M5, M6, M7, M8>
+{
+  typedef Synchronizer<LatestTime> Sync;
+  typedef PolicyBase<M0, M1, M2, M3, M4, M5, M6, M7, M8> Super;
+  typedef typename Super::Messages Messages;
+  typedef typename Super::Signal Signal;
+  typedef typename Super::Events Events;
+  typedef typename Super::RealTypeCount RealTypeCount;
+
+  LatestTime(uint32_t)
+  : parent_(0),
+    ros_clock_(RCL_ROS_TIME)
+  {
+  }
+
+  LatestTime(const LatestTime& e)
+  {
+    *this = e;
+  }
+
+  LatestTime& operator=(const LatestTime& rhs)
+  {
+    parent_ = rhs.parent_;
+    events_ = rhs.events_;
+    rates_ = rhs.rates_;
+    ros_clock_ = rhs.ros_clock_;
+
+    return *this;
+  }
+
+  void initParent(Sync* parent)
+  {
+    parent_ = parent;
+  }
+
+  template<int i>
+  void add(const typename std::tuple_element<i, Events>::type& evt)
+  {
+    RCUTILS_ASSERT(parent_);
+
+    std::lock_guard<std::mutex> lock(data_mutex_);
+
+    if(!received_msg<i>())
+    {
+      rates_.push_back(Rate(ros_clock_.now()));
+      // wait until we get each message once to publish
+      // then wait until we got each message twice to compute rates
+      // NOTE: this will drop a few messages of the faster topics until
+      //       we get one of the slowest so we can sync
+      std::get<i>(events_) = evt;  // adding here ensures we see even the slowest
+                                   // message twice before computing rate
+      return;
+    }
+
+    std::get<i>(events_) = evt;
+
+    rclcpp::Time now = ros_clock_.now();
+    rates_[i].compute_hz(now);
+    if(i == find_pivot(now) && is_full())
+    {
+      publish();
+    }
+  }
+
+private:
+  // assumed data_mutex_ is locked
+  void publish()
+  {
+    parent_->signal(std::get<0>(events_), std::get<1>(events_), std::get<2>(events_),
+                    std::get<3>(events_), std::get<4>(events_), std::get<5>(events_),
+                    std::get<6>(events_), std::get<7>(events_), std::get<8>(events_));
+  }
+
+  struct Rate
+  {
+    rclcpp::Time prev;
+    double hz{0.0};
+    double error{0.0};
+    bool do_hz_init{true};
+    bool do_error_init{true};
+    Rate(const rclcpp::Time &start)
+      : prev(start)
+    {
+    }
+
+    bool operator>(const Rate &that) const
+    {
+      return this->hz > that.hz;
+    }
+
+    void compute_hz(const rclcpp::Time &now)
+    {
+      double period = (now-prev).seconds();
+      if (do_hz_init) {
+        hz = 1.0/period;
+        do_hz_init = false;
+      } else {
+        if (do_error_init) {
+          error = fabs(hz - 1.0/period);
+          do_error_init = false;
+        } else {
+          if (fabs(hz - 1.0/period) > 10.0*error) {
+            // detected step change in rate so reset
+            do_hz_init = true;
+            do_error_init = true;
+            compute_hz(now);
+            return;
+          }
+          error = 0.7*error + 0.3*fabs(hz - 1.0/period);
+        }
+        hz = 0.9/period + 0.1*hz;
```

Comment:
Similarly, where do these magic numbers come from, and should they be configurable?

Reply:
also configs